flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/AggSqlFunction.scala

```diff
@@ -0,0 +1,177 @@
+    null
+  ) {
+
+  def getFunction: AggregateFunction[_, _] = aggregateFunction
+}
+
+object AggSqlFunction {
+
+  def apply(
+      name: String,
+      aggregateFunction: AggregateFunction[_, _],
+      returnType: TypeInformation[_],
+      typeFactory: FlinkTypeFactory): AggSqlFunction = {
+
+    new AggSqlFunction(name, aggregateFunction, returnType, typeFactory)
+  }
+
+  private[flink] def createOperandTypeInference(
+      aggregateFunction: AggregateFunction[_, _],
+      typeFactory: FlinkTypeFactory)
+  : SqlOperandTypeInference = {
+    /**
+      * Operand type inference based on [[AggregateFunction]] given information.
+      */
+    new SqlOperandTypeInference {
+      override def inferOperandTypes(
+          callBinding: SqlCallBinding,
+          returnType: RelDataType,
+          operandTypes: Array[RelDataType]): Unit = {
+
+        val operandTypeInfo = getOperandTypeInfo(callBinding)
+
+        val foundSignature = getAccumulateMethodSignature(aggregateFunction, operandTypeInfo)
+          .getOrElse(throw new ValidationException(s"Operand types of could not be inferred."))
+
+        val inferredTypes = getParameterTypes(aggregateFunction, foundSignature.drop(1))
+          .map(typeFactory.createTypeFromTypeInfo)
+
+        for (i <- operandTypes.indices) {
+          if (i < inferredTypes.length - 1) {
+            operandTypes(i) = inferredTypes(i)
+          } else if (null != inferredTypes.last.getComponentType) {
+            // last argument is a collection, the array type
+            operandTypes(i) = inferredTypes.last.getComponentType
+          } else {
+            operandTypes(i) = inferredTypes.last
+          }
+        }
+      }
+    }
+  }
+
+  private[flink] def createReturnTypeInference(
+      resultType: TypeInformation[_],
+      typeFactory: FlinkTypeFactory)
+  : SqlReturnTypeInference = {
+
+    new SqlReturnTypeInference {
+      override def inferReturnType(opBinding: SqlOperatorBinding): RelDataType = {
+        typeFactory.createTypeFromTypeInfo(resultType)
+      }
+    }
+  }
+
+  private[flink] def createOperandTypeChecker(aggregateFunction: AggregateFunction[_, _])
+  : SqlOperandTypeChecker = {
+
+    val signatures = getMethodSignatures(aggregateFunction, "accumulate")
+
+    /**
+      * Operand type checker based on [[AggregateFunction]] given information.
+      */
+    new SqlOperandTypeChecker {
+      override def getAllowedSignatures(op: SqlOperator, opName: String): String = {
+        s"$opName[${signaturesToString(aggregateFunction, "accumulate")}]"
+      }
+
+      override def getOperandCountRange: SqlOperandCountRange = {
+        var min = 255
+        var max = -1
+        signatures.foreach(
+          sig => {
+            val inputSig = sig.drop(1)
+            //do not count accumulator as input
+            var len = inputSig.length
+            if (len > 0 && inputSig(inputSig.length - 1).isArray) {
+              max = 254 // according to JVM spec 4.3.3
```

Comment:
We want to make sure number of parameters of all codeGened functions within the [min, max], otherwise it will trigger the JVM failure. For accumulate and retract method, the max (for Operand which does not take into account the accumulator) here should be 253, as function pointer itself takes one, and accumulator takes one. Yes, we should consider to add test cases for huge amount of inputs and Array case.